### ISSUE TITLE
libutp build issues

### ISSUE THREAD

**Issue by jpalus** [transmission/transmission#4748]
### What is the issue?

When updating to `transmission 4.0.0` I've hit a few issues with `libutp`:
- trying to build embedded `libutp` with `gcc 12.2.0` fails with:
```
/home/users/builder/rpm/BUILD/transmission-4.0.0/build/third-party/utp/src/utp/utp_internal.cpp:3394:21: error: this statement may fall through [-Werror=implicit-fallthrough=]
```
- if I disable `libutp` with `-DENABLE_UTP=OFF` I get:
```
CMake Error at libtransmission/CMakeLists.txt:267 (target_link_libraries):
  Target "transmission" links to:

    libutp::libutp

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.
```
- trying to use system `libutp` from https://github.com/bittorrent/libutp gives another build failure:
```
/home/users/builder/rpm/BUILD/transmission-4.0.0/libtransmission/peer-io.cc: In lambda function:
/home/users/builder/rpm/BUILD/transmission-4.0.0/libtransmission/peer-io.cc:733:40: error: 'struct utp_callback_arguments' has no member named 'u1'
/home/users/builder/rpm/BUILD/transmission-4.0.0/libtransmission/peer-io.cc: In lambda function:
/home/users/builder/rpm/BUILD/transmission-4.0.0/libtransmission/peer-io.cc:746:63: error: 'struct utp_callback_arguments' has no member named 'u1'
/home/users/builder/rpm/BUILD/transmission-4.0.0/libtransmission/peer-io.cc: In lambda function:
/home/users/builder/rpm/BUILD/transmission-4.0.0/libtransmission/peer-io.cc:758:47: error: 'struct utp_callback_arguments' has no member named 'u1'
make[2]: *** [libtransmission/CMakeFiles/transmission.dir/build.make:429: libtransmission/CMakeFiles/transmission.dir/peer-io.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/home/users/builder/rpm/BUILD/transmission-4.0.0/libtransmission/tr-utp.cc: In function 'uint64 {anonymous}::utp_callback(utp_callback_arguments*)':
/home/users/builder/rpm/BUILD/transmission-4.0.0/libtransmission/tr-utp.cc:137:58: error: 'struct utp_callback_arguments' has no member named 'u1'
/home/users/builder/rpm/BUILD/transmission-4.0.0/libtransmission/tr-utp.cc:137:76: error: 'struct utp_callback_arguments' has no member named 'u2'
```

So it seems currently using system `libutp` is not possible (`USE_SYSTEM_UTP`), `libutp` is not optional (`ENABLE_UTP`) and build files for embedded `libutp` need tweaking so `LIBUTP_ENABLE_WERROR` does not get set to `ON` while keeping `LIBUTP_STANDALONE_BUILD` to `ON` since it's required for install target.

### Which application of Transmission?

None

### Which version of Transmission?

4.0.0

**Comment by ckerr** [comment]
tl;dr in terms of unblocking you I'd say go into `third-party/libutp/CMakeLists.txt` and just turn off the Werror option

```diff
-  option(LIBUTP_ENABLE_WERROR "${PROJECT_NAME}: Treat warnings as errors" ${LIBUTP_STANDALONE_BUILD})
+  option(LIBUTP_ENABLE_WERROR "${PROJECT_NAME}: Treat warnings as errors" OFF)
```

I see three possible workitems for Transmission here:

1. Your guess that the system `libutp` is incompatible is almost certainly correct. `libutp` has some semver-major changes but the upstream project doesn't do releases AFAIK and Transmission can't just run pkg-config to see what version is preinstalled. It's likely _not_ detecting the version correctly; Transmission should probably sniff `utp.h` for a version string and give you a cleaner error message.

2. I don't understand why your build is enabling -Werror; it looks like somehow `LIBUTP_STANDALONE_BUILD` is set to ON which it shouldn't be unless you're building libutp in its own cmake directory. But anyway that CMakeLists.txt file comes from The Transmission project and I'm not sure why we ever enable -Werror when building for Transmission. (@mikedld might know?)

3. Maybe the `WITH_UTP` option to compile without UTP isn't needed anymore. It was added IIRC a decade ago when one distro had reservations about libutp but I think the library has proven itself safe since then. Also, the overwhelming majority of BitTorrent clients use UTP and so users who don't have it would be at a disadvantage.

For bookkeeping's sake I'm going to open separate tickets for those three issues, and I want to focus here in this ticket on getting it to build for you. @jpalus does disabling `LIBUTP_STANDALONE_BUILD` ubnlock you?

**Comment by jpalus** [comment]
> it looks like somehow LIBUTP_STANDALONE_BUILD is set to ON which it shouldn't be unless you're building libutp in its own cmake directory.
> does disabling LIBUTP_STANDALONE_BUILD ubnlock you?

Actually that was my very first attempt to drop the logic for setting `LIBUTP_STANDALONE_BUILD` to `ON` which results in yet another failure:
```
make[3]: *** No rule to make target 'install'.  Stop.
make[2]: *** [CMakeFiles/utp.dir/build.make:107: third-party/utp/src/utp-stamp/utp-install] Error 2
```

I guess that's because `LIBUTP_STANDALONE_BUILD` is used for `LIBUTP_ENABLE_INSTALL` as well:
https://github.com/transmission/libutp/blob/bf695bdfb047cdca9710ea9cffc4018669cf9548/CMakeLists.txt#L9

and the latter guards:
https://github.com/transmission/libutp/blob/bf695bdfb047cdca9710ea9cffc4018669cf9548/CMakeLists.txt#L92-L103

**Comment by ckerr** [comment]
OK. @mikedld wrote those and I'd rather give him a chance to weigh in instead of me second-guessing why that code is there.

In terms of just making this build for you, maybe remove the actual `-Werror` injection call to `add_compile_options()` at https://github.com/transmission/libutp/blob/bf695bdfb047cdca9710ea9cffc4018669cf9548/CMakeLists.txt#L36 ?

**Comment by jpalus** [comment]
Just to be clear I was able to compile `transmission 4.0.0` successfully after patching embedded `libutp` cmake file and forcing `LIBUTP_ENABLE_WERROR` to `OFF`.

**Comment by ckerr** [comment]
Awesome, I'm glad you got unblocked!

Since I've filed separate tickets for those workitems above (4750, 4751, 4752) I'm going to close this issue as completed. In particular 4752 would be the one for anyone else who hits this issue.
